### ISSUE TITLE
Redis close or keepalive

### ISSUE THREAD

**Issue by Roycohen** [openresty/lua-resty-redis#147]
Hi There,

We are using the module for a long time and I think we have a mem leak or something regarding the connections.
I would like to know the following:
1. Do we need to close redis connections? What happen if we are not closing the connection?
2. When to use close and when to use keepalive? Do we need to use keepalive all the time?
3. If we have timeout in connection do we still need to close or use keepalive on the object?
4. If we get timeout when making a command, what we need to do? Is there a chance for mem leak?

Basically I want to understand what may cause mem leaks and what action to take under several circumstances.

**Comment by spacewander** [comment]
Another option to detect leaking in C land: you could build OpenResty with ASAN support, like: https://github.com/openresty/openresty-packaging/blob/master/rpm/SPECS/openresty-openssl-asan.spec#L52, and see if there is any error report in the error log.

**Comment by terryburton** [comment]
> When you want to reuse the connection, use keepalive. If the connection is unreusable (for example, there is an network error happened in the connection). use close.

Is there a consistent way to know whether a return results in the connection being unusable such that is should be closed?

In particular:

* Does the returned error always indicate the connection state, somehow?
* Does the connection state depend entirely on the function that was invoked that returned an error? I.e. All errors from this function result in a non-reusable connection.
* Do some functions return various errors that may variously indicate a failure but the connection remains healthy (e.g. a duplicate key) or a failure due-to / resulting-in a bad connection. I.e. You might need to categorise every possible error condition determine how the failure should be handled.
* If the latter is the case, is there a reliable error handler that can probe the status of a connection to determine whether the connection should be immediately closed or whether it is in a healthy state to be reused.

Thanks for entertaining 20 questions...

**Comment by spacewander** [comment]
@terryburton 
To answer all the questions, we need a new blog.

Short answer:
As the doc says: 
> A Redis "error reply" results in a false value and a string describing the error.

If the command call returns `false, "msg"`, then it is a redis protocol error. If the call returns `nil, "msg"`, it is connection error and you can close the connection.
